presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,39 +1594,197 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
         return sum * 1000;
     }
 
-    private static double computeSphericalExcess(Polygon polygon, int start, int end)
+    @SqlNullable
+    @Description("Returns TRUE if and only if no points of right lie in the exterior of left, and at least one point of the interior of left lies in the interior of right")
+    @ScalarFunction("ST_Contains")
+    @SqlType(BOOLEAN)
+    public static Boolean stSphericalContains(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice left, @SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        if (leftGeometry.isEmpty()) {
+            return false;
+        }
+
+        OGCGeometry rightGeometry = deserialize(right);
+        if (rightGeometry.isEmpty()) {
+            return false;
+        }
+
+        validateSphericalType("ST_Contains", leftGeometry, EnumSet.of(POLYGON));
+        validateSphericalType("ST_Contains", rightGeometry, EnumSet.of(POINT));
+
+        Polygon polygon = (Polygon) leftGeometry.getEsriGeometry();
+        Point point = (Point) rightGeometry.getEsriGeometry();
+
+        if (point.getY() == MAX_LATITUDE || point.getY() == MIN_LATITUDE) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "When applied to SphericalGeography inputs, ST_Contains only supports points which are not poles.");
+        }
+
+        if (SphericalCalculator.containsPole(polygon)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "When applied to SphericalGeography inputs, ST_Contains only supports polygons which do not enclose poles.");
+        }
+
+        return contains(polygon, point);
+    }
+
+    private static boolean liesInRange(double point, double rangeStart, double rangeEnd)
     {
-        // Our calculations rely on not processing the same point twice
-        if (polygon.getPoint(end - 1).equals(polygon.getPoint(start))) {
-            end = end - 1;
+        // strictly exclusive
+        return Double.compare(point, rangeStart) > 0 && Double.compare(point, rangeEnd) < 0;
+    }
+
+    private static boolean haveSameLongitude(double x, double y)
+    {
+        if (Double.compare(x, y) == 0) {
+            return true;
         }
+        if ((Double.compare(x, 180.0) == 0 || Double.compare(y, 180.0) == 0) &&
+                (Double.compare(x, -180.0) == 0 && Double.compare(y, -180.0) == 0)) {
```

Comment:
This will always evaluate as false.  The first nested expression means at least one of `x` and `y` is `180.0`; the second nested expression requires both `x` and `y` to be `-180.0`.

A clearer way to write this that does what we want is:
```
if ( (x == 180.0 && y = -180.0) || (x == -180.0 && y == 180.0) ) {
    return true;
}
```